nautobot/extras/models/datasources.py

```diff
@@ -167,3 +174,55 @@ def sync(self, user, dry_run=False):
         if dry_run:
             return enqueue_git_repository_diff_origin_and_local(self, user)
         return enqueue_pull_git_repository_and_refresh_data(self, user)
+
+    @contextmanager
+    def clone_to_directory(self, path=None, branch=None, head=None, shallow=True):
+        """
+        Perform a (shallow) clone of the Git repository in a temporary directory.
+
+        Args:
+            path (str, optional): The absolute directory path to clone into. If not specified, `tempfile.gettempdir()` will be used.
+            branch (str, optional): The branch to checkout. If not set, the GitRepository.branch will be used.
+            head (str, optional): Git commit hash to check out instead of pulling branch latest.
+            shallow (bool, optional): If True, perform a shallow clone. Defaults to True.
+
+        Returns:
+            Returns the absolute path of the cloned repo if clone was successful, otherwise returns None.
+        """
+
+        try:
+            if branch and head:
+                raise ValueError("Cannot specify both branch and head")
+
+            try:
+                path_name = tempfile.mkdtemp(dir=path, prefix=self.slug)
+            except PermissionError as e:
+                logger.error(f"Failed to create temporary directory at {path}: {e}")
+                raise e
+
+            if not branch:
+                branch = self.branch
+
+            try:
+                repo_helper = GitRepo(path_name, self.remote_url, shallow=shallow)
+                head, _ = repo_helper.checkout(branch, head)
```

Comment:
https://github.blog/open-source/git/get-up-to-speed-with-partial-clone-and-shallow-clone/ seems like a useful reference. 

If I'm understanding it right, we should *not* generally support `clone_to_directory(head=...., shallow=True)` since by definition `shallow=True` implies "latest commits only".

Additionally, we should probably enhance `GitRepo.__init__()` to also take `branch` as an optional parameter so that we can do `git clone ... --depth=1 --branch=<branch>` rather than cloning the default branch and then trying to switch to the specified branch.